pkg/proxy/metrics/metrics.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// used in the middleware below to capture http status code.
+type statusCatcher struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (s *statusCatcher) WriteHeader(statusCode int) {
+	s.statusCode = statusCode
+	s.ResponseWriter.WriteHeader(statusCode)
+}
+
+// WithRequestCounter counts the number of requests that have passed through
+// the handler it wraps.
+func WithRequestCounter(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		delegate.ServeHTTP(sc, r)
+
+		code := strconv.Itoa(sc.statusCode)
+		verb := strings.ToUpper(r.Method)
+		requestCounter.WithContext(r.Context()).WithLabelValues(verb, code).Inc()
+	})
+}
+
+// WithLatencyTracking tracks the number of seconds it took the wrapped handler
+// to complete.
+func WithLatencyTracking(delegate http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		sc := &statusCatcher{
+			ResponseWriter: w,
+		}
+
+		start := time.Now()
+		delegate.ServeHTTP(sc, r)
+		elapsedSeconds := time.Since(start).Seconds()
+
+		code := strconv.Itoa(sc.statusCode)
+		verb := strings.ToUpper(r.Method)
+		requestLatencies.WithContext(r.Context()).WithLabelValues(verb, code).Observe(elapsedSeconds)
+	})
+}
+
+// TODO(csams): enhance these metrics to include shard url
+var (
+	requestCounter = compbasemetrics.NewCounterVec(
+		&compbasemetrics.CounterOpts{
+			Name:           "proxy_request_total",
```

Comment:
nit: naming convention would suggest `proxy_request_count` - but we actually get the counts for free from the histogram below, so we don't need this metric.

Reply:
Sure, I can change it if you like. I just slightly modified https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/metrics/metrics.go#L75-L83 and kept the convention there.

Reply:
With the histogram it's 100% superfluous.

Reply:
That's fair. We can just use the +Inf bucket. I'll update the PR.